### ISSUE TITLE
ext-ds Set not supported by the new `filter` filter

### ISSUE THREAD

**Issue by keksa** [twigphp/Twig#3015]
We use `ext-ds` and since the `if` condition on `for` was deprecated in Twig 2.10.0, we moved to `filter` tag with arrow functions.

When trying to use the filter on `Ds\Set`, it crashes with:
> CallbackFilterIterator::__construct() must implement interface Iterator, instance of Ds\Set given

This is the (simplified) template code:
```twig
{% for error in errors|filter(error => constant('ERRORS', item)[error] is defined) %}
    {{ constant('ERRORS', item)[error]|trans }}
{% endfor %}
```
The `errors` variable is instance of `Ds\Set`.

Since this was working before, I think that either `Ds\Set` should be supported or even anything implementing `\Traversable`.

**Comment by enumag** [comment]
It's not just a matter of Ds\Set but of any internal PHP structure that implements Traversable. See this note from PHP [docs](https://www.php.net/manual/en/class.traversable.php):

```
Internal (built-in) classes that implement this interface can be used in a foreach construct and do not need to implement IteratorAggregate or Iterator. 
```

**Comment by fabpot** [comment]
You can also use:

```twig
{% for error in errors %}
    {% if constant('ERRORS', item)[error] is defined %}
        {{ constant('ERRORS', item)[error]|trans }}
    {% endif %}
{% endfor %}
```

**Comment by keksa** [comment]
@fabpot I know there are workarounds, either the one you posted, or calling 'Set::toArray()', but that's not the point of my issue.
IMO there is no reason, why `filter` should not support `Traversable` interface, since that is, what is effectively supported by twig's `for` loop.

**Comment by fabpot** [comment]
@keksa Do you want to work on a pull request to add support for each Traversable objects?

**Comment by keksa** [comment]
@fabpot I can look into it, if you'll accept such PR.

**Comment by fabpot** [comment]
fixed by #3017
